test/org/zaproxy/zap/extension/websocket/client/ServerConnectionEstablisherUnitTest.java

```diff
@@ -0,0 +1,109 @@
+import org.parosproxy.paros.network.HttpMessage;
+import org.zaproxy.zap.extension.websocket.ExtensionWebSocket;
+import org.zaproxy.zap.extension.websocket.WebSocketProxy;
+import org.zaproxy.zap.extension.websocket.utility.WebSocketUtils;
+import org.zaproxy.zap.testutils.WebSocketTestUtils;
+import org.zaproxy.zap.testutils.websocket.server.NanoWebSocketConnection;
+import org.zaproxy.zap.testutils.websocket.server.NanoWebSocketTestServer;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+import java.util.Stack;
+
+public class ServerConnectionEstablisherUnitTest extends WebSocketTestUtils {
+	private final static int WEBSOCKET_PORT = 8885;
+	private final static int TIMEOUT = 100000;
+	private final static String HOST_NAME = "localhost";
+	
+	@Before
+	public void openWebSocketServer() throws Exception {
+		super.startWebSocketServer(HOST_NAME,WEBSOCKET_PORT,TIMEOUT);
+		super.setUpZap();
+	}
+	
+	@Override
+	protected void setUpMessages() {
+		mockMessages(new ExtensionWebSocket());
+	}
+	
+	@Test
+	public void shouldReceiveUpgradeStatusCode() throws Exception {
+		ServerConnectionEstablisher establisher = new ServerConnectionEstablisher();
+		HttpMessage handshakeRequest = new HttpMessage(HttpHandshakeBuilder.getHttpHandshakeRequestHeader(new URI("http://" + HOST_NAME + ":" + WEBSOCKET_PORT,true)));
+		NanoWebSocketTestServer webSocketServer = super.getWebSocketTestServer();
+		System.out.println(handshakeRequest.getRequestHeader().toString());
+		establisher.send(new HandshakeConfig(handshakeRequest,false,false));
+		assertEquals(101,handshakeRequest.getResponseHeader().getStatusCode());
+		assertEquals(webSocketServer.getLastConnection()
+						.getHandshakeResponse().getHeader(HttpHandshakeBuilder.SEC_WEB_SOCKET_ACCEPT)
+				, WebSocketUtils.encodeWebSocketKey(handshakeRequest.getRequestHeader().getHeader(HttpHandshakeBuilder.SEC_WEB_SOCKET_KEY)));
+		
+	}
+	
+	@Test
+	public void shouldReturnWSProxy() throws Exception {
+		ServerConnectionEstablisher establisher = new ServerConnectionEstablisher();
+		HttpMessage handshakeRequest = new HttpMessage(HttpHandshakeBuilder.getHttpHandshakeRequestHeader(new URI("http://" + HOST_NAME + ":" + WEBSOCKET_PORT,true)));
+		
+		WebSocketProxy webSocketProxy = establisher.send(new HandshakeConfig(handshakeRequest,false,false));
+		NanoWebSocketConnection webSocketConnection = super.getWebSocketTestServer().getLastConnection();
+		assertEquals(101,handshakeRequest.getResponseHeader().getStatusCode());
+		assertNotNull(webSocketProxy);
+		assertTrue(webSocketProxy.isConnected());
+		assertEquals(webSocketConnection.getHandshakeResponse().getHeader(HttpHandshakeBuilder.SEC_WEB_SOCKET_ACCEPT)
+				,WebSocketUtils.encodeWebSocketKey(handshakeRequest.getRequestHeader().getHeader(HttpHandshakeBuilder.SEC_WEB_SOCKET_KEY)));
+	}
+	
+	@Test
+	public void shouldReceiveMessagesFromServer() throws Exception {
```

Comment:
This should be moved to the other class, it's mainly testing the `WebSocketProxy`.